primitives/session/src/lib.rs

```diff
@@ -106,6 +108,76 @@ impl GetValidatorCount for MembershipProof {
 	}
 }
 
+/// A trait for online node inspection in a session.
+///
+/// This trait is used to decouple the pallet-session dependency from im-online
+/// module so that the user of im-online & offences modules can pass any list of
+/// validators that are considered to be online in each session, particularly useful
+/// for the Substrate-based projects having their own staking implementation
+/// instead of using pallet-staking directly.
+pub trait ValidatorSet<AccountId> {
+	/// Type for representing validator id in a session.
+	///
+	/// This could be `frame_support::Parameter`, but such import is not allowed in primitives.
+	type ValidatorId: codec::FullCodec + Clone + Eq + sp_std::fmt::Debug;
+	/// A type for converting `AccountId` to `ValidatorId`.
+	type ValidatorIdOf: Convert<AccountId, Option<Self::ValidatorId>>;
+
+	/// Returns current session index.
+	fn session_index() -> SessionIndex;
+
+	/// Returns all the validators ought to be online in a session.
+	///
+	/// The returned validators are all expected to be running an authority node.
+	fn validators() -> Vec<Self::ValidatorId>;
+}
+
+/// `ValidatorSet` combined with identification type for pallet-session-historical module.
+pub trait ValidatorSetWithIdentification<AccountId>: ValidatorSet<AccountId> {
+	/// Full identification of `ValidatorId`.
+	type Identification: codec::FullCodec + Clone + Eq + sp_std::fmt::Debug;
+	/// A type for converting `ValidatorId` to `Identification`.
+	type IdentificationOf: Convert<Self::ValidatorId, Option<Self::Identification>>;
+}
+
+/// A session handler for specific key type.
+pub trait OneSessionHandler<ValidatorId>: BoundToRuntimeAppPublic {
```

Comment:
All these 3 traits here are frame specific. Please move them to `frame-support`. They do not belong here.

Reply:
@bkchr Please have another look, hopefully, I have applied all the suggestions, let me know if I miss anything.